Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/OpenShiftController.cs

```diff
@@ -370,6 +345,225 @@ internal async Task ProcessOpenShiftsAsync(string isRequestFromLogicApp)
+                var openShiftsInCacheToProcess = identicalOpenShifts.Where(x => x.KronosOpenShiftUniqueId == openShiftHash).ToList();
+                var kronosOpenShiftsToProcess = openShiftsFoundList.Where(x => x.KronosUniqueId == openShiftHash).ToList();
+
+                // Calculate the difference in number of open shifts between Kronos and cache
+                var cacheOpenShiftSlotCount = 0;
+                openShiftsInCacheToProcess.ForEach(x => cacheOpenShiftSlotCount += x.KronosSlots);
+                var numberOfOpenShiftsToRemove = cacheOpenShiftSlotCount - kronosOpenShiftsToProcess.Count;
+
+                if (numberOfOpenShiftsToRemove > 0)
+                {
+                    // More open shifts in cache than found in Kronos
+                    await this.RemoveAdditionalOpenShiftsFromTeamsAsync(allRequiredConfigurations, mappedOrgJobEntity, openShiftsInCacheToProcess, numberOfOpenShiftsToRemove).ConfigureAwait(false);
+                    continue;
+                }
+
+                if (numberOfOpenShiftsToRemove < 0)
+                {
+                    // Less open shifts in cache than found in Kronos
+                    var openShiftsToAdd = new List<OpenShiftRequestModel>();
+
+                    for (int i = numberOfOpenShiftsToRemove; i < 0; i++)
+                    {
+                        openShiftsToAdd.Add(kronosOpenShiftsToProcess.First());
+                    }
+
+                    await this.CreateEntryOpenShiftsEntityMappingAsync(allRequiredConfigurations, openShiftsToAdd, lookUpEntriesFoundList, monthPartitionKey, mappedOrgJobEntity).ConfigureAwait(false);
+                    continue;
+                }
+
+                // The number of open shifts in Kronos and Teams matches meaning no action is needed
+                continue;
+            }
+        }
+
+        /// <summary>
+        /// This method will remove open shifts from Teams in the event we find more OS in cache
+        /// than retrieved from Kronos.
+        /// </summary>
+        /// <param name="allRequiredConfigurations">The required configuration details.</param>
+        /// <param name="mappedOrgJobEntity">The team deatils.</param>
+        /// <param name="openShiftsToProcess">All of the matching entities in cache.</param>
+        /// <param name="numberOfOpenShiftsToRemove">The number of matching entities we want to remove.</param>
+        /// <returns>A unit of execution.</returns>
+        private async Task RemoveAdditionalOpenShiftsFromTeamsAsync(
+            SetupDetails allRequiredConfigurations,
+            TeamToDepartmentJobMappingEntity mappedOrgJobEntity,
+            List<AllOpenShiftMappingEntity> openShiftsToProcess,
+            int numberOfOpenShiftsToRemove)
+        {
+            while (numberOfOpenShiftsToRemove != 0)
+            {
+                if (!openShiftsToProcess.Any())

```

Comment:
validate this once at the top of the method rather than in every loop iteration and ensure that the count is >= numberOfOpenShiftsToRemove